src/json.hpp

```diff
@@ -466,16 +490,51 @@ caught.,other_error}
 class other_error : public exception
 {
   public:
-    static other_error create(int id_, const std::string& what_arg)
-    {
-        std::string w = exception::name("other_error", id_) + what_arg;
-        return other_error(id_, w.c_str());
-    }
+    other_error(int id_, const std::string& what_arg) : 
+      exception(id_, format_what(id_, what_arg).c_str()) {}
 
   private:
-    other_error(int id_, const char* what_arg) : exception(id_, what_arg) {}
+    static std::string format_what(int id_, const std::string& what_arg)
+    {
+        return exception::name("other_error", id_) + what_arg;
+    }
 };
 
+/*!
+@brief helper function to throw the a downcasted exception.
+*/
+template<class Exception>
+void try_throw_exception(exception* ex)
+{
+#ifdef JSON_EXCEPTIONS_ENABLED
+    // Only use RTTI if necessary (exceptions are enabled). This way
+    // can be compiled with no RTTI.
+    Exception* down = dynamic_cast<Exception*>(ex);
+    if(down != nullptr)
+    {
+        JSON_THROW(*down);
+    }
+#else
+    ex = ex; // suppress unreferenced formal parameter warnings
+    std::abort();
+#endif
+}
```

Comment:
I think this looks cleaner

    #ifdef JSON_EXCEPTIONS_ENABLED
    template<class Exception>
    void try_throw_exception(exception* ex)
    {
        // Only use RTTI if necessary (exceptions are enabled). This way
        // can be compiled with no RTTI.
        if(Exception* down = dynamic_cast<Exception*>(ex))
        {
            JSON_THROW(*down);
        }
    }
    #else
    template<class Exception>
    void try_throw_exception(exception* /*unused*/)
    {
        std::abort();
    }
    #endif

On the other hand, why is this even needed?  (It's definitely not if @theodelrieu 's suggested change is made, and I would say it's not even if it isn't, just throw `ex.get()`, no need for the downcasting.

Reply:
Agreed, that does look cleaner thanks. I'll try it out. As for just throwing without downcasting I couldn't get that to work. `catch (const SpecificError& e) {}` would never get hit.

Reply:
Yeah, I missed that you were throwing by value instead of by pointer, which unfortunately does require all this extra casting over throwing directly when the exception is created.  This delayed throw method does lose some of the call stack context, so it may be better to make the choice whether to throw or place in the unique_ptr<> at the point of the THROW macro.